### ISSUE TITLE
Decide on build tool

### ISSUE THREAD

**Issue by tliron** [nephio-project/nephio#113]
Some of us have default to using Make, but should we really?

* Pro: Make is familiar to many of us
* Con: Make is awful

Another possible option is to use straightforward bash scripts. We need to plug our build process into Prow, anyway, so it shouldn't matter if it's `make test` or `test.sh`.

**Comment by johnbelamaric** [comment]
Personally I prefer make. It's well known and well understood. I don't think it's so bad.

**Comment by henderiw** [comment]
I agree that make is well known and it is used by many people. if you know what it does/doesn't it is not that bad.

**Comment by gvbalaji** [comment]
Thanks @tliron . Personally I have no preference.  But  make is more ubiquitous and we have our repos and build system to work with it  . Is there any strong reason to change it?

**Comment by tliron** [comment]
1) Makefile is not much used in the Go world.
2) I think it could be a barrier of entry, another new language for potential contributers to learn.

**Comment by radoslawc** [comment]
In my opinion it makes little difference whether one uses (ba)sh script or Makefile as they are basically the same, though Makefile tab usage as delimiters occasionally causes some confusion. Another con for make is that an image in which prow job will run has to have make installed and if we're not using any make features, only just execute shell commands with parameters then I'd say we're better with simple shell script.

Bazel which seems what cool kids are using these days can do a lot of magic including reproducible builds or build docker images by itself, but it's enormous and because of otherwise nifty features of caching and sandboxing, one-off builds (which all CI code check jobs are, as pods get discarded after job finishes) take a lot of time. Also at build/test time it's depending on external repos (for example to pull go build templates). Literal 'hello world' takes around 20 seconds to build compared with less than one sec for 'go build'
```
bazel build //...  0.03s user 0.04s system 0% cpu 20.110 total
go build             0.29s user 0.35s system 78% cpu 0.814 total
```
Mage seems to be another popular solution, it's basically make but you write your build 'script' in go which is a pro. If we're considering barrier of entry, then with mage it's only GO that one would have to deal with.

So I guess the main question is what advanced features during the build process we need to justify anything more than putting together few commands in shell script.

**Comment by johnbelamaric** [comment]
K8s started with Bazel (because it's an OSS version of an internal Google tool), and ended up ripping it out. Which was a huge effort. It's great for the right use case but I don't see an advantage for us.

I wouldn't want to go with anything other than make or bash - a new tool like mage means we all need to learn something new. Make has been around for > 30 years and is pretty well established, I imagine 99%+ of people developing on Linux have used it...

**Comment by tliron** [comment]
In my opinion Make is overkill for any Go project. The `go` compilation process is so straightforward (single command does all) that we don't need any of Make's sophisticated (and complex) task management and optimization.

For that reason I'm in favor of just having bash scripts. We can establish some best practices and uniformity to those. (We would have to do the same for Make, too, should we want to continue using it.)
